src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1132,10 +1133,79 @@ private AuthenticationResult acquireTokenLocalCall(final CallbackHandler callbac
     private boolean promptUser(PromptBehavior prompt) {
         return prompt == PromptBehavior.Always || prompt == PromptBehavior.REFRESH_SESSION;
     }
+    
+    /**
+     * check the redirectUri before sending the request
+     * if the redirectUri from the client does not match the valid redirectUri
+     * the client app would not jump to the login page
+     * redirectUri format %PREFIX://%PACKAGE_NAME/%SIGNATURE
+     * 
+     * @param callbackHandle
+     * @param request
+     * @throws UnsupportedEncodingException
+     * @return true if the RedictUri is valid
+     */
+    private boolean verifyBrokerRedirectUri(final AuthenticationRequest request){   
+        final String methodName = ":verifyBrokerRedirectUri";
+        String errMsg = "";
+        final String inputUri = request.getRedirectUri();
+        AuthenticationException exception = null;
+        
+        if(StringExtensions.IsNullOrBlank(inputUri))
+        {
+            errMsg = "the redirectUri is null or blank";
+            exception = new AuthenticationException(ADALError.DEVELOPER_REDIRECTURI_INVALID, errMsg);
+            Logger.e(TAG + methodName, errMsg , "", ADALError.DEVELOPER_REDIRECTURI_INVALID);     
+        }
+        else if(!inputUri.startsWith(AuthenticationConstants.Broker.REDIRECT_PREFIX + "://"))
+        {
+            errMsg = "redirectUri does not start with the valid redict prefix " + AuthenticationConstants.Broker.REDIRECT_PREFIX; 
+            exception = new AuthenticationException(ADALError.DEVELOPER_REDIRECTURI_INVALID, errMsg);
+            Logger.e(TAG + methodName, errMsg , "", ADALError.DEVELOPER_REDIRECTURI_INVALID);     
+        } 
+        else
+        {
+            try 
+            {
+                PackageHelper packageHelper = new PackageHelper(mContext);
+                String base64URLEncodePackagename = URLEncoder.encode(mContext.getPackageName(), AuthenticationConstants.ENCODING_UTF8);
+                String base64URLEncodeSignature = URLEncoder.encode(packageHelper.getCurrentSignatureForPackage(mContext.getPackageName()), AuthenticationConstants.ENCODING_UTF8);
+                if(!inputUri.startsWith(AuthenticationConstants.Broker.REDIRECT_PREFIX + "://" + base64URLEncodePackagename + "/"))
+                {
+                    errMsg = "the package name of redirectUri is not as expected. The expected package name is " + base64URLEncodePackagename;
```

Comment:
Let's change this message to "The base64 url encoded package name component of the redirect uri does not match the expected value.  This apps package name is: <package_name> so the redirect uri is expected to be: <calculated_redirect_uri>"

We should try to make error messages as clear and actionable as possible.  If it's possible to create a message that tells the developer exactly what they need to do to fix the problem, then we should do so.